### PR TITLE
[WI-46976] Add stub map with its generation script

### PR DESCRIPTION
**Pull request by theofidry**
/cc @Ocramius

I am not sure how to commit the generation scrip as a pre-commit hook. If someone has an idea...

### CODE REVIEW

**Comment by Ocramius** [comment]
> as a pre-commit hook

Can't really: pre-commit hooks are not delivered together with the git repository.

**Comment by theofidry** [comment]
Having issues here since the script needs to be executed with PHP 7.2+ (it relies on `nikic/php-parser` - which is already a dev dependency)... Still trying to figure this out, but any help there would be appreciated

**Comment by Ocramius** [comment]
```
$ if [[ -n $(git status -s) ]] then;
  echo "The commited map is not up to date. Run 'php generate-stub-map.php' to generate the map and commit it.\n";
  exit 1;
fi
/home/travis/.travis/functions: eval: line 104: syntax error near unexpected token `then'
/home/travis/.travis/functions: eval: line 104: `if [[ -n $(git status -s) ]] then;'
```

I suggest putting this in an actual script, so that you can have a shebang, and therefore you know which beast you are throwing this at.

**Comment by MaXal** [comment]
Please apply the following patch to fix OOM that occurs because of the parsing of generated map in tests:
```
Index: tests/Parsers/StubParser.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- tests/Parsers/StubParser.php	(date 1560203868000)
+++ tests/Parsers/StubParser.php	(date 1561035713339)
@@ -52,7 +52,7 @@
             );
         /** @var SplFileInfo $file */
         foreach ($stubsIterator as $file) {
-            if (!$fileCondition($file) ||
+            if (!$fileCondition($file) || basename(dirname($file->getRealPath())) === "phpstorm-stubs" ||
                 strpos($file->getRealPath(), 'vendor') || strpos($file->getRealPath(), '.git') ||
                 strpos($file->getRealPath(), 'tests') || strpos($file->getRealPath(), '.idea')) {
                 continue;
Index: .travis.yml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- .travis.yml	(date 1560203868000)
+++ .travis.yml	(date 1561035772452)
@@ -16,7 +16,7 @@
   - php generate-stub-map.php
 script:
   - docker-compose -f docker-compose.yml pull
-  - docker-compose -f docker-compose.yml run php /opt/project/vendor/bin/phpunit /opt/project/tests/
+  - docker-compose -f docker-compose.yml run php /opt/project/phpstorm-stubs/vendor/bin/phpunit /opt/project/phpstorm-stubs/tests/
   - |
     if [[ -n $(git status -s) ]] then;
       echo "The commited map is not up to date. Run 'php generate-stub-map.php' to generate the map and commit it.\n";
`

**Comment by MaXal** [comment]
@theofidry I'm sorry I forgot to add `docker-compose.yml` to the patch.
Please change: `- .:/opt/project` => `- .:/opt/project/phpstorm-stubs`.

```
Index: docker-compose.yml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- docker-compose.yml	(date 1560203868000)
+++ docker-compose.yml	(date 1561392003036)
@@ -3,4 +3,4 @@
   php:
     image: php:7.3
     volumes:
-     - .:/opt/project
+     - .:/opt/project/phpstorm-stubs
```

**Comment by theofidry** [comment]
@MaXal I applied the patch but I'm not sure this is enough. I've allowed edit from maintainers in case you want to play around with it

File: PhpStormStubsMap.php [hunk 1]
```diff
@@ -5749,8 +5749,8 @@ final class PhpStormStubsMap
   'DB2_PASSTHRU' => 'ibm_db2/ibm_db2.php',
   'DB2_SCROLLABLE' => 'ibm_db2/ibm_db2.php',
   'DB2_XML' => 'ibm_db2/ibm_db2.php',
-  'DEBUG_BACKTRACE_IGNORE_ARGS' => 'Core/Core.php',
-  'DEBUG_BACKTRACE_PROVIDE_OBJECT' => 'Core/Core.php',
+  'DEBUG_BACKTRACE_IGNORE_ARGS' => 'Core/Core_d.php',
+  'DEBUG_BACKTRACE_PROVIDE_OBJECT' => 'Core/Core_d.php',
```
Comment (theofidry) [inline PhpStormStubsMap.php]: Just a thought but maybe the stub map dumper should throw when multiple files declare the same symbol?

Comment (MaXal) [reply PhpStormStubsMap.php]: @theofidry That is a good idea to implement! But I will add this check to the standard tests. Map tester should test map first of all!

**Comment by MaXal** [comment]
@theofidry Thank you for your request and opening it for modification! I hope this map will be helpful for the community!

**Comment by theofidry** [comment]
Most credits are due to @kukulich who originally implemented it for [BetterReflection](https://github.com/Roave/BetterReflection). Glad to see it landing here and I hope it will also help to improve the quality of the stubs!

**Pull request merged**